ocrd_tesserocr/recognize.py

```diff
@@ -19,62 +24,251 @@ def process(self):
+                regions = page.get_TextRegion()
+                if not regions:
+                    log.warn("Page contains no text regions") 
+                for region in regions:
+                    log.debug("Recognizing text in region '%s'", region.id)
+                    if maxlevel == 'region':
+                        region_xywh = xywh_from_points(region.get_Coords().points)
+                        tessapi.SetRectangle(region_xywh['x'], region_xywh['y'], region_xywh['w'], region_xywh['h'])
+                        tessapi.SetPageSegMode(PSM.SINGLE_BLOCK)
+                        if region.get_TextEquiv():
+                            log.warn("Region '%s' already contains text results", region.id)
+                        region.add_TextEquiv(TextEquivType(Unicode=tessapi.GetUTF8Text().rstrip("\n\f")))
+                        continue # next region (to avoid indentation below)
+                    ## line, word, or glyph level:
                     textlines = region.get_TextLine()
-                    log.info("About to recognize text in %i lines of region '%s'", len(textlines), region.id)
+                    if not textlines:
+                        log.warn("Region '%s' contains no text lines", region.id)
                     for line in textlines:
                         log.debug("Recognizing text in line '%s'", line.id)
-                        xywh = xywh_from_points(line.get_Coords().points)
-                        tessapi.SetRectangle(xywh['x'], xywh['y'], xywh['w'], xywh['h'])
-                        tessapi.SetPageSegMode(PSM.SINGLE_LINE)
-                        #  log.debug("xywh: %s", xywh)
-                        line.add_TextEquiv(TextEquivType(Unicode=tessapi.GetUTF8Text()))
+                        line_xywh = xywh_from_points(line.get_Coords().points)
+                        #  log.debug("xywh: %s", line_xywh)
+                        tessapi.SetRectangle(line_xywh['x'], line_xywh['y'], line_xywh['w'], line_xywh['h'])
+                        tessapi.SetPageSegMode(PSM.SINGLE_LINE) # RAW_LINE fails with Tesseract 3 models and is worse with Tesseract 4 models
+                        if line.get_TextEquiv():
+                            log.warn("Line '%s' already contains text results", line.id)
                         #  tessapi.G
-                        #  print(tessapi.AllWordConfidences())
-                        if self.parameter['textequiv_level'] == 'glyph':
-                            for word in line.get_Word():
-                                log.debug("Recognizing text in word '%s'", word.id)
-                                xywh = xywh_from_points(word.get_Coords().points)
-                                tessapi.SetRectangle(xywh['x'], xywh['y'], xywh['w'], xywh['h'])
+                        line_conf = 1.0
+                        for (word_no, conf) in enumerate(tessapi.AllWordConfidences()): # or tessapi.MeanTextConf()?
+                            line_conf *= conf/100.0
```

Comment:
Is that specified somewhere, that line confidence is the product of word confidences?

Reply:
No, not really. For true probabilities that would be correct I think. But apart from the percentage scale, these numbers need not sum up to 1 (or 100, respectively). This is what I see with Tesseract 3 models. For LSTM models it looks more like a probability, but that may change soon. So shall we prefer `MeanTextConf`?